python/ray/train/examples/jax_mnist_example.py

```diff
@@ -0,0 +1,206 @@
+
+    @functools.partial(jax.pmap, axis_name="ensemble")
+    def apply_model(state, images, labels):
+        """Computes gradients, loss and accuracy for a single batch."""
+
+        def loss_fn(params):
+            logits = MLP().apply({"params": params}, images)
+            one_hot = jax.nn.one_hot(labels, 10)
+            loss = jnp.mean(optax.softmax_cross_entropy(logits=logits, labels=one_hot))
+            return loss, logits
+
+        grad_fn = jax.value_and_grad(loss_fn, has_aux=True)
+        (loss, logits), grads = grad_fn(state.params)
+        accuracy = jnp.mean(jnp.argmax(logits, -1) == labels)
+        return grads, loss, accuracy
+
+    @jax.pmap
+    def update_model(state, grads):
+        return state.apply_gradients(grads=grads)
+
+    @functools.partial(jax.pmap, axis_name="ensemble")
+    def train_step(state, images, labels):
+        """Computes gradients, loss and accuracy for a single batch."""
+
+        def loss_fn(params):
+            logits = MLP().apply({"params": params}, images)
+            one_hot = jax.nn.one_hot(labels, 10)
+            loss = jnp.mean(optax.softmax_cross_entropy(logits=logits, labels=one_hot))
+            return loss, logits
+
+        grad_fn = jax.value_and_grad(loss_fn, has_aux=True)
+        (loss, logits), grads = grad_fn(state.params)
+        grads = lax.pmean(grads, axis_name="ensemble")
+        accuracy = jnp.mean(jnp.argmax(logits, -1) == labels)
+        state = state.apply_gradients(grads=grads)
+        loss = lax.pmean(loss, axis_name="ensemble")
+        accuracy = lax.pmean(accuracy, axis_name="ensemble")
+        return state, loss, accuracy
+
+
+    def train_epoch(state, train_ds, batch_size, rng):
+        """Train for a single epoch."""
+        train_ds_size = len(train_ds["image"])
+        steps_per_epoch = train_ds_size // batch_size
+
+        epoch_loss = []
+        epoch_accuracy = []
+
+        for i in range(steps_per_epoch):
+            batch_images = train_ds["image"][i * batch_size : (i + 1) * batch_size]
+            batch_labels = train_ds["label"][i * batch_size : (i + 1) * batch_size]
+            batch_images = shard(batch_images)
+            batch_labels = shard(batch_labels)
+            state, loss, accuracy = train_step(state, batch_images, batch_labels)
+            epoch_loss.append(jax_utils.unreplicate(loss))
+            epoch_accuracy.append(jax_utils.unreplicate(accuracy))
+        train_loss = np.mean(epoch_loss)
+        train_accuracy = np.mean(epoch_accuracy)
+        return state, train_loss, train_accuracy
+
+
+    """Execute model training"""
+    # get the configuration
+    learning_rate = config['learning_rate']
+    momentum = config['momentum']
+    batch_size = config['batch_size']
+    num_epochs = config['num_epochs']
+    worker_batch_size = batch_size // train.world_size()
+
+    # Create datasets
+    train_ds, test_ds = get_datasets()
```

Comment:
Is it possible to use ray datasets instead?

Reply:
Yes, see #25385